src/memory-analyzer/analyze_symbol.cpp

```diff
@@ -0,0 +1,476 @@
+    deref_pointer, zero_initializer(type, location, ns), type, location);
+
+  declaration.operands().resize(2);
+  declaration.op1() = target_expr;
+  return declaration;
+}
+
+exprt symbol_analyzert::fill_array_with_values(
+  const std::string &symbol_id,
+  exprt &array,
+  const typet &type,
+  const source_locationt &location)
+{
+  for(size_t i = 0; i < array.operands().size(); ++i)
+  {
+    const std::string cell_access = symbol_id + "[" + std::to_string(i) + "]";
+    const typet target_type = ns.follow(array.operands()[i].type());
+    if(is_c_char_pointer(target_type))
+    {
+      const std::string memory_location = gdb_api.get_memory(cell_access);
+      const symbol_exprt char_ptr(cell_access, target_type);
+      array.operands()[i] =
+        declare_and_initalize_char_ptr(char_ptr, memory_location, location);
+    }
+    else
+    {
+      array.operands()[i] = fill_expr_with_values(
+        cell_access, array.operands()[i], target_type, location);
+    }
+  }
+  return array;
+}
+
+exprt symbol_analyzert::fill_expr_with_values(
+  const std::string &symbol_id,
+  exprt expr,
+  const typet &type,
+  const source_locationt &location)
+{
+  if(expr.is_constant() && (is_c_int_derivate(type) || is_c_char(type)))
+  {
+    const std::string value = gdb_api.get_value(symbol_id);
+    const mp_integer int_rep = string2integer(value, DECIMAL_SYSTEM);
+    return from_integer(int_rep, type);
+  }
+  else if(expr.is_constant() && is_c_bool(type))
+  {
+    const std::string value = gdb_api.get_value(symbol_id);
+    return from_c_boolean_value(value, type);
+  }
+  else if(expr.is_constant() && is_c_enum(expr.type()))
+  {
+    const std::string value = gdb_api.get_value(symbol_id);
+    return convert_memeber_name_to_enum_value(
+      value, to_c_enum_type(expr.type()));
+  }
+  else if(is_struct(type))
+  {
+    return fill_struct_with_values(symbol_id, expr, type, location);
+  }
+  else if(is_array(type))
+  {
+    expr = fill_array_with_values(symbol_id, expr, type, location);
+  }
+  else
+  {
+    throw symbol_analysis_exceptiont(
+      "unexpected thing:  " + expr.pretty() + "\nexception end\n");
+  }
+  return expr;
+}
+
+exprt symbol_analyzert::fill_char_struct_member_with_values(
+  const symbol_exprt &field_access,
+  const exprt &default_expr,
+  const source_locationt &location)
+{
+  const std::string memory_location =
+    gdb_api.get_memory(id2string(field_access.get_identifier()));
+  if(memory_location != "0x0")
+  {
+    return declare_and_initalize_char_ptr(
+      field_access, memory_location, location);
+  }
+  return default_expr;
+}
+
+exprt symbol_analyzert::fill_struct_with_values(
+  const std::string &symbol_id,
+  exprt &expr,
+  const typet &type,
+  const source_locationt &location)
+{
+  const struct_typet structt = to_struct_type(type);
+  for(size_t i = 0; i < expr.operands().size(); ++i)
+  {
+    exprt &operand = expr.operands()[i];
+    const typet &resolved_type = ns.follow(operand.type());
```

Comment:
Shouldn't be done here, only resolve were actually needed (i.e., with the is-it-a-struct-type below)